six/modules/c++/six.sicd/source/AreaPlaneUtility.cpp

```diff
@@ -0,0 +1,260 @@
+        (static_cast<sys::SSize_T>(data.imageData->numRows) ==
+        data.imageData->fullImage.row) &&
+        (static_cast<sys::SSize_T>(data.imageData->numCols) ==
+        data.imageData->fullImage.col))
+    {
+        std::vector<types::RowCol<sys::SSize_T> > validPixelRegion(
+                data.imageData->validData.size());
+        for (size_t ii = 0; ii < validPixelRegion.size(); ++ii)
+        {
+            validPixelRegion[ii] = data.imageData->validData[ii];
+        }
+
+        std::vector<types::RowCol<sys::SSize_T> > hullPoints;
+        math::ConvexHull<sys::SSize_T> convexHull(validPixelRegion, hullPoints);
+
+        // ConvexHull makes a copy of the first vertex into the last,  so pop
+        // off the last entry
+        cornersPix.resize(hullPoints.size() - 1);
+        for (size_t ii = 0; ii < hullPoints.size() - 1; ++ii)
+        {
+            cornersPix[ii] = hullPoints[ii];
+        }
+    }
+    else
+    {
+        // Use the grid corners
+        cornersPix.push_back(RowColDouble(0.0, 0.0));
+        cornersPix.push_back(RowColDouble(
+                0.0, data.getNumCols() - 1));
+        cornersPix.push_back(RowColDouble(
+                data.getNumRows() - 1, data.getNumCols() - 1));
+        cornersPix.push_back(RowColDouble(
+                data.getNumRows() - 1, 0.0));
+    }
+
+    return cornersPix;
+}
+
+types::RowCol<Vector3> AreaPlaneUtility::deriveUnitVectors(
+        const ComplexData& data)
+{
+    std::auto_ptr<scene::SceneGeometry> geometry(
+            Utilities::getSceneGeometry(&data));
+
+    types::RowCol<Vector3> unitVectors;
+    const Vector3 slantPlaneX = geometry->getSlantPlaneX();
+    const Vector3 normal = geometry->getGroundPlaneNormal();
+    unitVectors.row = slantPlaneX - normal * slantPlaneX.dot(normal);
+    unitVectors.row.normalize();
+    unitVectors.row *= -1;
+
+    unitVectors.col = math::linear::cross(normal, unitVectors.row);
+    return unitVectors;
+}
+
+RowColDouble AreaPlaneUtility::deriveSpacing(
+        const ComplexData& data,
+        const types::RowCol<Vector3>& unitVectors)
+{
+    std::auto_ptr<scene::SceneGeometry> geometry(
+            Utilities::getSceneGeometry(&data));
+    geometry->setImageVectors(unitVectors.row, unitVectors.col);
+    geometry->setOutputPlaneVectors(unitVectors.row, unitVectors.col);
+
+    const RowColDouble origResolution(
+            data.grid->row->impulseResponseWidth,
+            data.grid->col->impulseResponseWidth);
+    const RowColDouble bandwidth(
+            data.grid->row->impulseResponseBandwidth,
+            data.grid->col->impulseResponseBandwidth);
+    const RowColDouble k = (origResolution * bandwidth) /
+            Constants::EXCESS_BANDWIDTH_FACTOR;
+    const types::RgAz<double> resolution(origResolution.row / k.row,
+            origResolution.col / k.col);
+    const RowColDouble groundResolution =
+            geometry->getGroundResolution(resolution);
+    const double SAMPLE_DENSITY = 1.5;
```

Comment:
We should let them pass this in too into the public functions (but default it to 1.5)... add it as a static default to the class so the number doesn't have to be copied in two spots.  It'll be really unusual someone wants to change this so put it after the segmentList bool flag